DIPs/DIP1044.md

```diff
@@ -76,9 +76,7 @@ it will no longer link because the mangling will be different. But for `extern (
 safety is not part of the mangling and it will compile and link without error. This has always
 relied on the user getting it correct.
 
-Interfaces to extern C and C++ functions should always be explicitly marked.
```

Comment:
I was going to write such a DIP myself; I was waiting for `-preview=dip1000` to become the default. I think that any and all declarations with no body should have explicit `@safe`/`@trusted`/`@system` attributes and that even considering linkage is a mistake.

Reply:
How is that considered a mistake?

You are interfacing with a language that has no notion of any safety guarantees. Ignoring differences between languages and the interfaces between them is a mistake.

Reply:
Because `extern(C)` does *not* mean "this is C code", it means "C linkage" and could very well be `@safe` D code.

Reply:
The problem is, a `@safe` `extern(C)` function that is not defined in-place is not verified by the compiler – and `@safe` means that the compiler checked it (up to calls to `@trusted`). I’m no expert on this, but as far as I understand it’s about mangling. Attributes are not part of the signature of `extern(C)` functions. `extern(C) void f() @safe;` could be defined elsewhere as `extern(C) void f() @system { … }`. I think D needs *some* solution to this; Maybe I’m naïve, but it could be as easy as generating mock symbols that actually include attributes that lead to linker errors when the attributes don’t match.

At least until 2.096.0, this worked:
```D
import a;
extern(C) void f() @safe;
void main() @safe { f(); }
```
```D
module a;
extern(C) void f() @system { int* p;  int a;  p = &a; }
```
If you change `extern(C)` to `extern(D)`, you get the link error.

Reply:
> Because `extern(C)` does _not_ mean "this is C code", it means "C linkage" and could very well be `@safe` D code.

If it's `@safe` D code then they can just import it. It doesn't matter if it's C code, or linkage, it is still going by C's rules. Just cause safe D code can have C linkage doesn't mean it should expose a **C** linkage of safe.

Reply:
Importing does not safeguard you from calling a `@safe` annotated `extern(C)` _declared_ function (e.g. from a .di file) whose _definition_ (implementation in a .d file) is annotated `@system`. That’s because the annotation is not visible for the linker.
I guess the easiest way would be to allow `extern(C) @safe` function declarations only if they’re also definitions. The proper way would be for D to do something that safeguards you at least in simple and obvious cases.